File: src/d1_bridge_node.cpp
#include <array>
#include <atomic>
#include <cmath>
#include <cstdio>
#include <csignal>
#include <memory>
#include <mutex>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joint_state.hpp"

#include <unitree/robot/channel/channel_publisher.hpp>
#include <unitree/robot/channel/channel_subscriber.hpp>

#include "msg/ArmString_.hpp"
#include "msg/PubServoInfo_.hpp"

using namespace unitree::robot;

static constexpr const char* DDS_SERVO_TOPIC = "current_servo_angle";
static constexpr const char* DDS_CMD_TOPIC   = "rt/arm_Command";
static constexpr const char* ROS_STATE_TOPIC = "/d1/joint_states";
static constexpr const char* ROS_CMD_TOPIC   = "/d1/joint_commands";

// Names match the IsaacLab training env joint names exactly.
// Joint1-Joint6 = arm (servo0-servo5), Joint_L/Joint_R = gripper (both driven by servo6).
static constexpr std::array<const char*, 8> JOINT_NAMES = {
    "Joint1", "Joint2", "Joint3", "Joint4", "Joint5", "Joint6", "Joint_L", "Joint_R"
};

class D1BridgeNode : public rclcpp::Node {
public:
    D1BridgeNode() : Node("d1_bridge_node"), seq_(0), has_prev_(false) {
        joint_state_pub_ = create_publisher<sensor_msgs::msg::JointState>(ROS_STATE_TOPIC, 10);

        joint_cmd_sub_ = create_subscription<sensor_msgs::msg::JointState>(
            ROS_CMD_TOPIC, 10,
            [this](sensor_msgs::msg::JointState::SharedPtr msg) { on_joint_command(msg); });

        arm_pub_ = std::make_shared<ChannelPublisher<unitree_arm::msg::dds_::ArmString_>>(DDS_CMD_TOPIC);
        arm_pub_->InitChannel();

        // DDS callback runs in a separate CycloneDDS thread
        servo_sub_ = std::make_shared<ChannelSubscriber<unitree_arm::msg::dds_::PubServoInfo_>>(DDS_SERVO_TOPIC);
        servo_sub_->InitChannel([this](const void* msg) { on_servo_feedback(msg); });

        RCLCPP_INFO(get_logger(), "D1 bridge ready. Listening on DDS '%s'.", DDS_SERVO_TOPIC);
    }

private:
    void on_servo_feedback(const void* raw) {
        if (!active_.load(std::memory_order_relaxed)) return;
        const auto* pm = static_cast<const unitree_arm::msg::dds_::PubServoInfo_*>(raw);

        // 7 hardware servo channels → 8 sim joints.
        // servo0-5: Joint1-Joint6 (arm). servo6: gripper, split symmetrically to Joint_L/Joint_R.
        std::array<double, 8> pos_deg = {
            pm->servo0_data_(), pm->servo1_data_(), pm->servo2_data_(),
            pm->servo3_data_(), pm->servo4_data_(), pm->servo5_data_(),
            pm->servo6_data_(),   // Joint_L
            pm->servo6_data_()    // Joint_R (same channel)
        };

        auto now = get_clock()->now();
        std::array<double, 8> vel_rads{};
        {
            std::lock_guard<std::mutex> lock(state_mutex_);
            if (has_prev_) {
                double dt = (now - prev_time_).seconds();
                if (dt > 1e-6) {
                    for (int i = 0; i < 8; ++i)
                        vel_rads[i] = (pos_deg[i] - prev_pos_deg_[i]) * M_PI / 180.0 / dt;
                }
            }
            prev_pos_deg_ = pos_deg;
            prev_time_    = now;
            has_prev_     = true;
        }

        auto msg = sensor_msgs::msg::JointState();
        msg.header.stamp = now;
        for (int i = 0; i < 8; ++i) {
            msg.name.push_back(JOINT_NAMES[i]);
            msg.position.push_back(pos_deg[i] * M_PI / 180.0);
            msg.velocity.push_back(vel_rads[i]);
        }
        joint_state_pub_->publish(msg);
    }

    void on_joint_command(const sensor_msgs::msg::JointState::SharedPtr msg) {
        if (msg->position.size() < 7) {
            RCLCPP_WARN(get_logger(), "Joint command has %zu positions, expected 7 — ignoring.",
                        msg->position.size());
            return;
        }

        // Policy outputs 8 joint positions in radians (Joint1-6, Joint_L, Joint_R).
        // D1 JSON expects 7 angles in degrees; gripper channel (angle6) = average of Joint_L/R.
        auto deg = [](double rad) { return rad * 180.0 / M_PI; };
        double gripper_deg = (deg(msg->position[6]) + deg(msg->position[7])) * 0.5;

        char json[512];
        std::snprintf(json, sizeof(json),
            "{\"seq\":%d,\"address\":1,\"funcode\":2,\"data\":"
            "{\"mode\":1"
            ",\"angle0\":%.3f,\"angle1\":%.3f,\"angle2\":%.3f"
            ",\"angle3\":%.3f,\"angle4\":%.3f,\"angle5\":%.3f,\"angle6\":%.3f}}",
            seq_++,
            deg(msg->position[0]), deg(msg->position[1]), deg(msg->position[2]),
            deg(msg->position[3]), deg(msg->position[4]), deg(msg->position[5]),
            gripper_deg);

        unitree_arm::msg::dds_::ArmString_ dds_msg{};
        dds_msg.data_() = json;
        arm_pub_->Write(dds_msg);
    }

    rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr joint_state_pub_;
    rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr joint_cmd_sub_;

    std::shared_ptr<ChannelPublisher<unitree_arm::msg::dds_::ArmString_>> arm_pub_;
    std::shared_ptr<ChannelSubscriber<unitree_arm::msg::dds_::PubServoInfo_>> servo_sub_;

    std::atomic<bool> active_{true};
    std::mutex state_mutex_;
    std::array<double, 8> prev_pos_deg_{};
    rclcpp::Time prev_time_{0, 0, RCL_ROS_TIME};
    bool has_prev_;
    int seq_;
};

int main(int argc, char* argv[]) {
    // Must initialize Unitree DDS before rclcpp::init() — both use CycloneDDS and
    // ChannelFactory::Init() is not idempotent when called after rclcpp already owns the domain.
    ChannelFactory::Instance()->Init(0, "enx0c3796c78061");
    rclcpp::init(argc, argv);
    // rclcpp's default SIGINT/SIGTERM handler calls rclcpp::shutdown(), which triggers
    // CycloneDDS teardown that conflicts with the Unitree SDK's participant and corrupts
    // the heap. Override it AFTER rclcpp::init() to just _exit() instead.
    signal(SIGINT,  [](int) { _exit(0); });
    signal(SIGTERM, [](int) { _exit(0); });

    rclcpp::spin(std::make_shared<D1BridgeNode>());
    _exit(0);
}

File: src/d1_bridge_node.py
#!/usr/bin/env python3
# Requires: pip install cyclonedds
# Requires: unitree_sdk2_python in PYTHONPATH or at the path below
import math
import os
import signal
import sys
import threading

sys.path.insert(0, '/home/master26/unitree_sdk2_python')

from dataclasses import dataclass
import cyclonedds.idl as idl
import cyclonedds.idl.annotations as annotate
import cyclonedds.idl.types as types

from unitree_sdk2py.core.channel import (
    ChannelPublisher, ChannelSubscriber, ChannelFactoryInitialize,
)

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState


NETWORK_INTERFACE = 'enx0c3796c78061'
DDS_SERVO_TOPIC   = 'current_servo_angle'
DDS_CMD_TOPIC     = 'rt/arm_Command'
ROS_STATE_TOPIC   = '/d1/joint_states'
ROS_CMD_TOPIC     = '/d1/joint_commands'
JOINT_NAMES = [
    'Joint_0', 'Joint_1', 'Joint_2', 'Joint_3',
    'Joint_4', 'Joint_5', 'Joint_6',
]


@dataclass
@annotate.final
@annotate.autoid("sequential")
class ArmString_(idl.IdlStruct, typename="unitree_arm::msg::dds_::ArmString_"):
    data_: str = ""


@dataclass
@annotate.final
@annotate.autoid("sequential")
class PubServoInfo_(idl.IdlStruct, typename="unitree_arm::msg::dds_::PubServoInfo_"):
    servo0_data_: types.float32 = 0.0
    servo1_data_: types.float32 = 0.0
    servo2_data_: types.float32 = 0.0
    servo3_data_: types.float32 = 0.0
    servo4_data_: types.float32 = 0.0
    servo5_data_: types.float32 = 0.0
    servo6_data_: types.float32 = 0.0


class D1BridgeNode(Node):
    def __init__(self):
        super().__init__('d1_bridge_node')
        self._seq = 0
        self._lock = threading.Lock()
        self._has_prev = False
        self._prev_pos_deg = [0.0] * 7
        self._prev_time_sec = 0.0

        self._state_pub = self.create_publisher(JointState, ROS_STATE_TOPIC, 10)
        self.create_subscription(JointState, ROS_CMD_TOPIC, self._on_joint_command, 10)

        # DDS publisher to the arm firmware
        self._arm_pub = ChannelPublisher(DDS_CMD_TOPIC, ArmString_)
        self._arm_pub.Init()

        # DDS subscriber for servo feedback — callback runs in a CycloneDDS thread
        self._servo_sub = ChannelSubscriber(DDS_SERVO_TOPIC, PubServoInfo_)
        self._servo_sub.Init(self._on_servo_feedback)

        self._feedback_sub = ChannelSubscriber('arm_Feedback', ArmString_)
        self._feedback_sub.Init(lambda msg: self.get_logger().info(f'arm_Feedback: {msg.data_}'))

        self.get_logger().info(f"D1 bridge ready. Listening on DDS '{DDS_SERVO_TOPIC}'.")

    def _on_servo_feedback(self, pm: PubServoInfo_):
        pos_deg = [
            pm.servo0_data_, pm.servo1_data_, pm.servo2_data_,
            pm.servo3_data_, pm.servo4_data_, pm.servo5_data_,
            pm.servo6_data_,
        ]

        now = self.get_clock().now()
        now_sec = now.nanoseconds * 1e-9
        vel_rads = [0.0] * 7

        with self._lock:
            if self._has_prev:
                dt = now_sec - self._prev_time_sec
                if dt > 1e-6:
                    for i in range(7):
                        vel_rads[i] = (pos_deg[i] - self._prev_pos_deg[i]) * math.pi / 180.0 / dt
            self._prev_pos_deg = pos_deg[:]
            self._prev_time_sec = now_sec
            self._has_prev = True

        msg = JointState()
        msg.header.stamp = now.to_msg()
        msg.name = list(JOINT_NAMES)
        msg.position = [p * math.pi / 180.0 for p in pos_deg]
        msg.velocity = vel_rads
        self._state_pub.publish(msg)

    def _on_joint_command(self, msg: JointState):
        if len(msg.position) < 7:
            self.get_logger().warn(
                f'Joint command has {len(msg.position)} positions, expected 7 — ignoring.'
            )
            return

        def deg(rad): return rad * 180.0 / math.pi

        seq = self._seq
        self._seq += 1

        json_str = (
            f'{{"seq":{seq},"address":1,"funcode":2,"data":{{'
            f'"mode":1'
            f',"angle0":{deg(msg.position[0]):.3f}'
            f',"angle1":{deg(msg.position[1]):.3f}'
            f',"angle2":{deg(msg.position[2]):.3f}'
            f',"angle3":{deg(msg.position[3]):.3f}'
            f',"angle4":{deg(msg.position[4]):.3f}'
            f',"angle5":{deg(msg.position[5]):.3f}'
            f',"angle6":{deg(msg.position[6]):.3f}'
            f'}}}}'
        )
        self._arm_pub.Write(ArmString_(data_=json_str))


def main(args=None):
    # Must initialize Unitree DDS before rclpy.init() — both use CycloneDDS and
    # ChannelFactory.Init() is not idempotent once rclpy already owns the domain.
    ChannelFactoryInitialize(0, NETWORK_INTERFACE)

    rclpy.init(args=args)

    # Create the node (and all DDS entities) before setting signal handlers —
    # CycloneDDS reinstalls its own C-level SIGINT handler during DataWriter/DataReader
    # creation and would override any handler set before this point.
    node = D1BridgeNode()

    signal.signal(signal.SIGINT,  lambda *_: os._exit(0))
    signal.signal(signal.SIGTERM, lambda *_: os._exit(0))

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    os._exit(0)


if __name__ == '__main__':
    main()
